File: workouts_udec_frontend/src/types/workout.ts
import type { Exercise } from './exercise';

export const WorkoutStatus = {
  ACTIVE: "ACTIVE",
  COMPLETED: "COMPLETED",
  CANCELLED: "CANCELLED"
} as const;

export type WorkoutStatus = typeof WorkoutStatus[keyof typeof WorkoutStatus];

export interface Workout {
  id: number;
  user_id: number;
  name?: string;
  started_at: string;
  completed_at?: string;
  notes?: string;
  template_id?: number;
  workout_exercises?: WorkoutExercise[];
}

export interface WorkoutCreate {
  name?: string;
  template_id?: number;
  notes?: string;
}

export interface WorkoutUpdate {
  name?: string;
  notes?: string;
  completed_at?: string;
}

export interface WorkoutExercise {
  id: number;
  workout_id: number;
  exercise_id: number;
  exercise: Exercise;
  order_index: number;
  notes?: string;
  sets?: ExerciseSet[];
}

export interface WorkoutExerciseCreate {
  exercise_id: number;
  order_index: number;
  notes?: string;
}

export interface WorkoutExerciseUpdate {
  order_index?: number;
  notes?: string;
}

export interface ExerciseSet {
  id: number;
  workout_exercise_id: number;
  set_number: number;
  reps?: number;
  weight?: number;
  duration?: number;
  rest_duration?: number;
  notes?: string;
  completed: boolean;
  created_at: string;
  updated_at: string;
}

export interface ExerciseSetCreate {
  set_number: number;
  reps?: number;
  weight?: number;
  duration?: number;
  rest_duration?: number;
  notes?: string;
  completed?: boolean;
}

export interface ExerciseSetUpdate {
  reps?: number;
  weight?: number;
  duration?: number;
  rest_duration?: number;
  notes?: string;
  completed?: boolean;
}

export interface WorkoutTemplate {
  id: number;
  name: string;
  description?: string;
  is_public: boolean;
  created_by: number;
  created_at: string;
  updated_at: string;
  template_exercises?: WorkoutTemplateExercise[];
}

export interface WorkoutTemplateExerciseCreate {
  exercise_id: number;
  order_index: number;
  suggested_sets?: number;
  suggested_reps?: number;
  suggested_weight?: number;
  suggested_duration?: number;
}

export interface WorkoutTemplateCreate {
  name: string;
  description?: string;
  is_public?: boolean;
  exercises?: WorkoutTemplateExerciseCreate[];
}

export interface WorkoutTemplateUpdate {
  name?: string;
  description?: string;
  is_public?: boolean;
}

export interface WorkoutTemplateExercise {
  id: number;
  template_id: number;
  exercise_id: number;
  exercise?: Exercise;
  order_index: number;
  suggested_sets?: number;
  suggested_reps?: number;
  suggested_weight?: number;
  suggested_duration?: number;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface WorkoutTemplateExerciseCreate {
  exercise_id: number;
  order_index: number;
  suggested_sets?: number;
  suggested_reps?: number;
  suggested_weight?: number;
  suggested_duration?: number;
}

export interface WorkoutSummary {
  workout: Workout;
  total_exercises: number;
  total_sets: number;
  total_volume?: number;
  total_duration_minutes?: number;
}

export interface ExerciseProgression {
  exercise: Exercise;
  sessions: {
    date: string;
    best_set: ExerciseSet;
    total_volume?: number;
    total_reps?: number;
  }[];
  personal_records: {
    max_weight?: number;
    max_reps?: number;
    max_duration?: number;
    max_volume?: number;
  };
}

File: workouts_udec_backend/app/schemas/workout.py
"""Schemas Pydantic para plantillas de entrenamiento, ejercicios y sets.

Contiene los modelos usados por la API para crear, actualizar y listar workouts
y plantillas.
"""

from typing import List, Optional
from pydantic import BaseModel
from datetime import datetime
from app.schemas.exercise import Exercise


class WorkoutTemplateExerciseBase(BaseModel):
    """Modelo base para ejercicios dentro de una plantilla (sugerencias)."""
    exercise_id: int
    order_index: int
    suggested_sets: Optional[int] = None
    suggested_reps: Optional[int] = None
    suggested_weight: Optional[float] = None
    suggested_duration: Optional[int] = None


class WorkoutTemplateExerciseCreate(WorkoutTemplateExerciseBase):
    """Schema para añadir un ejercicio a una plantilla."""
    pass


class WorkoutTemplateExercise(WorkoutTemplateExerciseBase):
    """Representación completa de un ejercicio dentro de una plantilla."""
    id: int
    template_id: int
    exercise: Exercise

    class Config:
        from_attributes = True


class WorkoutTemplateBase(BaseModel):
    """Campos básicos de una plantilla de entrenamiento."""
    name: str
    description: Optional[str] = None
    is_public: bool = False


class WorkoutTemplateCreate(WorkoutTemplateBase):
    """Schema para crear una plantilla incluyendo sus ejercicios."""
    exercises: List[WorkoutTemplateExerciseCreate] = []


class WorkoutTemplateUpdate(BaseModel):
    """Schema para actualizar una plantilla de entrenamiento."""
    name: Optional[str] = None
    description: Optional[str] = None
    is_public: Optional[bool] = None


class WorkoutTemplate(WorkoutTemplateBase):
    """Plantilla completa con metadatos y ejercicios asociados."""
    id: int
    created_by: int
    created_at: datetime
    updated_at: Optional[datetime] = None
    template_exercises: List[WorkoutTemplateExercise] = []

    class Config:
        from_attributes = True


class ExerciseSetBase(BaseModel):
    """Modelo base para un set dentro de un ejercicio."""
    set_number: int
    reps: Optional[int] = None
    weight: Optional[float] = None
    duration: Optional[int] = None
    rest_duration: Optional[int] = None
    completed: bool = False


class ExerciseSetCreate(ExerciseSetBase):
    """Schema para crear un set en un ejercicio durante el workout."""
    pass


class ExerciseSetUpdate(BaseModel):
    """Schema para actualizar un set existente."""
    set_number: Optional[int] = None
    reps: Optional[int] = None
    weight: Optional[float] = None
    duration: Optional[int] = None
    rest_duration: Optional[int] = None
    completed: Optional[bool] = None


class ExerciseSet(ExerciseSetBase):
    """Representación de un set ya persistido (con id)."""
    id: int

    class Config:
        from_attributes = True


class WorkoutExerciseBase(BaseModel):
    """Campos básicos para un ejercicio dentro de un workout."""
    exercise_id: int
    order_index: int
    notes: Optional[str] = None


class WorkoutExerciseCreate(WorkoutExerciseBase):
    """Schema para crear un ejercicio dentro de un workout (incluye sets)."""
    sets: List[ExerciseSetCreate] = []


# Need to import Exercise schema to avoid circular imports
from app.schemas.exercise import Exercise


class WorkoutExercise(WorkoutExerciseBase):
    """Ejercicio asociado a un workout con sus sets y metadatos."""
    id: int
    workout_id: int
    exercise: Exercise
    sets: List[ExerciseSet] = []

    class Config:
        from_attributes = True


class WorkoutBase(BaseModel):
    """Campos básicos de un workout (sesión de entrenamiento)."""
    name: Optional[str] = None
    notes: Optional[str] = None
    template_id: Optional[int] = None


class WorkoutCreate(WorkoutBase):
    """Schema para crear un workout completo con ejercicios."""
    exercises: List[WorkoutExerciseCreate] = []


class WorkoutUpdate(BaseModel):
    """Schema para actualizar metadatos de un workout."""
    name: Optional[str] = None
    notes: Optional[str] = None
    completed_at: Optional[datetime] = None


class WorkoutHistory(BaseModel):
    """Datos de resumen de workouts completados para historiales."""
    id: int
    user_id: int
    name: Optional[str] = None
    started_at: datetime
    completed_at: Optional[datetime] = None
    notes: Optional[str] = None
    template_id: Optional[int] = None

    class Config:
        from_attributes = True


class Workout(WorkoutBase):
    """Representación completa de un workout con ejercicios y metadatos."""
    id: int
    user_id: int
    started_at: datetime
    completed_at: Optional[datetime] = None
    workout_exercises: List[WorkoutExercise] = []

    class Config:
        from_attributes = True
